pkg/reconciler/v1alpha1/broker/broker.go

```diff
@@ -0,0 +1,442 @@
+
+	if b.DeletionTimestamp != nil {
+		// Everything is cleaned up by the garbage collector.
+		return reconcile.Result{}, nil
+	}
+
+	c, err := r.reconcileChannel(ctx, b)
+	if err != nil {
+		logging.FromContext(ctx).Error("Problem reconciling the channel", zap.Error(err))
+		b.Status.MarkChannelFailed(err)
+		return reconcile.Result{}, err
+	} else if c.Status.Address.Hostname == "" {
+		logging.FromContext(ctx).Info("Channel is not yet ready", zap.Any("c", c))
+		// Give the Channel some time to get its address. One second was chosen arbitrarily.
+		return reconcile.Result{RequeueAfter: time.Second}, nil
+	}
+	b.Status.MarkChannelReady()
+
+	_, err = r.reconcileFilterDeployment(ctx, b)
+	if err != nil {
+		logging.FromContext(ctx).Error("Problem reconciling filter Deployment", zap.Error(err))
+		return reconcile.Result{}, err
+	}
+	_, err = r.reconcileFilterService(ctx, b)
+	if err != nil {
+		logging.FromContext(ctx).Error("Problem reconciling filter Service", zap.Error(err))
+		return reconcile.Result{}, err
+	}
+	b.Status.MarkFilterReady()
+
+	_, err = r.reconcileIngressDeployment(ctx, b, c)
+	if err != nil {
+		logging.FromContext(ctx).Error("Problem reconciling ingress Deployment", zap.Error(err))
+		return reconcile.Result{}, err
+	}
+
+	svc, err := r.reconcileIngressService(ctx, b)
+	if err != nil {
+		logging.FromContext(ctx).Error("Problem reconciling ingress Service", zap.Error(err))
+		return reconcile.Result{}, err
+	}
+	b.Status.MarkIngressReady()
+	b.Status.SetAddress(names.ServiceHostName(svc.Name, svc.Namespace))
+
+	return reconcile.Result{}, nil
+}
+
+// updateStatus may in fact update the broker's finalizers in addition to the status.
+func (r *reconciler) updateStatus(broker *v1alpha1.Broker) (*v1alpha1.Broker, error) {
+	ctx := context.TODO()
+	objectKey := client.ObjectKey{Namespace: broker.Namespace, Name: broker.Name}
+	latestBroker := &v1alpha1.Broker{}
+
+	if err := r.client.Get(ctx, objectKey, latestBroker); err != nil {
+		return nil, err
+	}
+
+	brokerChanged := false
+
+	if !equality.Semantic.DeepEqual(latestBroker.Finalizers, broker.Finalizers) {
+		latestBroker.SetFinalizers(broker.ObjectMeta.Finalizers)
+		if err := r.client.Update(ctx, latestBroker); err != nil {
+			return nil, err
+		}
+		brokerChanged = true
+	}
+
+	if equality.Semantic.DeepEqual(latestBroker.Status, broker.Status) {
+		return latestBroker, nil
+	}
+
+	if brokerChanged {
+		// Re-fetch.
+		latestBroker = &v1alpha1.Broker{}
+		if err := r.client.Get(ctx, objectKey, latestBroker); err != nil {
+			return nil, err
+		}
+	}
+
+	latestBroker.Status = broker.Status
+	if err := r.client.Status().Update(ctx, latestBroker); err != nil {
+		return nil, err
+	}
+
+	return latestBroker, nil
+}
+
+// reconcileFilterDeployment reconciles Broker's 'b' filter deployment.
+func (r *reconciler) reconcileFilterDeployment(ctx context.Context, b *v1alpha1.Broker) (*v1.Deployment, error) {
+	expected := resources.MakeFilterDeployment(&resources.FilterArgs{
```

Comment:
Would it be efficient to keep global/static variables for FilterDeployment and FilterService rather than making it each time?

Reply:
They differ based on the specific `Broker` they are being generated for, so we can't use global variables for them.